### PR TITLE
Feature/nginx ingress

### PR DESCRIPTION
**Pull request by shapirov103**
*Issue #70: Expand NGINX Addon*

*Description of changes:*

Implementation of external dns add-on and NGINX addon with AWS load balancer controller integration and route 53 integration as well as example of host-based routing with NGINX.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: docs/addons/nginx.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+```typescript
+import { NginxAddOn, ClusterAddOn, EksBlueprint }  from '@shapirov/cdk-eks-blueprint';
+
+const subdomain  = ...;
+const nginxAddOn = new NginxAddOn(({ 
+    internetFacing: true, 
+    backendProtocol: "tcp", 
+    externaDnsHostname: subdomain, 
+    crossZoneEnabled: false 
+    }));
```
Comment (askulkarni2) [inline docs/addons/nginx.md:20]: Minor: indentation

File: lib/addons/external-dns/index.ts [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+/**
+ * Direct import hosted zone provider, based on a known hosted zone ID. 
+ * Recommended method if hosted zone id is known, as it avoids extra look-ups.
+ */
+export class ImportHostedZoneProvider  implements HostedZoneProvider {
+
+    constructor(private hostedZoneId: string, private id?: string) { }
+
+    provide(clusterInfo: ClusterInfo): IHostedZone[] {
+        return [HostedZone.fromHostedZoneId(clusterInfo.cluster.stack, this.id ??  `${this.hostedZoneId}-Import`, 'ZOJJZC49E0EPZ')];
```
Comment (askulkarni2) [inline lib/addons/external-dns/index.ts:55]: Replace hard coded ID with provided one.

File: lib/addons/index.ts [hunk 3]
```diff
@@ -4,6 +4,11 @@ export { AwsLoadBalancerControllerAddOn} from './aws-loadbalancer-controller'
...
 export { ContainerInsightsAddOn } from './container-insights'
 export { ClusterAutoScalerAddOn } from './cluster-autoscaler'
+export { ExternalDnsAddon, LookupHostedZoneProvider, DelegatingHostedZoneProvider } from './external-dns'
 export { MetricsServerAddOn } from './metrics-server'
 export { NginxAddOn } from './nginx'
-export { XrayAddOn } from './xray'
\ No newline at end of file
+export { XrayAddOn } from './xray'
+
+export class Constants {
```
Comment (askulkarni2) [inline lib/addons/index.ts:12]: Should we consider a separate `constants.ts` file for versions and enums?

**Review by askulkarni2 — COMMENTED** [review]
Some minor comments, but otherwise lgtm!

File: lib/addons/external-dns/index.ts (hunk 2 above)
Comment (shapirov103) [reply lib/addons/external-dns/index.ts:55]: Good catch!

File: lib/addons/index.ts (hunk 3 above)
Comment (shapirov103) [reply lib/addons/index.ts:12]: I added constants for just add-ons at the moment. I suppose those could be different for each category - addons, cluster providers, teams, but I don't exclude we may require a global one in the future as well. Assuming the approach of constants per module is viable, unless we have a case when it does not work.

File: lib/addons/external-dns/index.ts (hunk 2 above)
Comment (shapirov103) [reply lib/addons/external-dns/index.ts]: Addressed

File: docs/addons/nginx.md (hunk 1 above)
Comment (shapirov103) [reply docs/addons/nginx.md]: Addressed

File: docs/addons/external-dns.md [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+  type: LoadBalancer
+  ports:
+  - port: 80
+    protocol: TCP
+    targetPort: 80
+  selector:
+    name: your-app
+```
+
+Note the `external-dns.alpha.kubernetes.io/hostname` annotation for the service name that allows specifying its domain name.
```
Comment (kcoleman731) [inline docs/addons/external-dns.md:63]: and what will this do?

File: docs/addons/external-dns.md [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    name: your-app
+```
+
+Note the `external-dns.alpha.kubernetes.io/hostname` annotation for the service name that allows specifying its domain name. 
+
+## Hosted Zone Providers
+
+In order for external DNS to work, you need to supply one or more hosted zones. You need a reference to the EKS stack in order to either create or look up the hosted zones. 
+
+To help customers handle common use cases for Route 53 provisioning the platform provides a few convenience providers.
```
Comment (kcoleman731) [inline docs/addons/external-dns.md:69]: the framework.

File: docs/addons/external-dns.md [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        hostedZone: new ssp.addons.LookupHostedZoneProvider(myDomainName);
+    })
+```
+
+If the hosted zone ID is known, then the recommended approach is to use a `ImportHostedZoneProvider` which bypasses any lookup calls.
+
+```typescript
+    const myHostedZoneId = ""
+    new ssp.addons.ExternalDnsAddon({
+        hostedZone: new ssp.addons.ImportHostedZoneProvider(myDomainName);
```
Comment (kcoleman731) [inline docs/addons/external-dns.md:86]: should this be in the `addons` namespace? Feels more like a `util`.

File: docs/addons/nginx.md [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+NGINX add-on is a standard SSP add-on and it is usage is similar to the rest of the add-ons.
+
+```typescript
+import { NginxAddOn, ClusterAddOn, EksBlueprint }  from '@shapirov/cdk-eks-blueprint';
+
+const subdomain  = ...;
+const nginxAddOn = new NginxAddOn(({ 
+    internetFacing: true, 
+    backendProtocol: "tcp",
```
Comment (kcoleman731) [inline docs/addons/nginx.md:17]: are these default values? If yes, lets hide in docs?

File: docs/addons/nginx.md [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+To validate that installation is successful run the following command:
+
+```bash
+$ kubectl get po -n kube-system
+NAME                                                              READY   STATUS    RESTARTS   AGE
+ssp-addon-nginx-ingress-78b8567p4q6   1/1     Running   0          4d10h
+```
+
+Note that the ingress is deployed in the `kube-system` namespace.
```
Comment (kcoleman731) [inline docs/addons/nginx.md:41]: which ingress?

File: docs/addons/nginx.md [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+Once deployed, it allows applications to create ingress objects and use host based routing with external DNS support, if External DNS Add-on is installed.
+
+## Configuration
+
+ - `backendProtocol`: indication for AWS Load Balancer controller with respect to the protocol supported on the load balancer. TCP by default.
```
Comment (kcoleman731) [inline docs/addons/nginx.md:47]: should we make these tables? You have different formatting for configuration in the ExternalDNSAddon

File: docs/addons/nginx.md [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+
+ - `crossZoneEnabled`: whether to create a cross-zone load balancer with the service that backs NGINX.
+
+ - `internetFacing`: whether the created load balancer is internet facing. Defaults to `true` if not specified. Internal load balancer is provisioned if set to `false`
+
+ -  `targetType`: `IP` or `instance` mode. Defaults to `IP` which requires VPC-CNI and has better performance eliminating a hop through kubeproxy. Instance mode leverages traditional NodePort mode on the instances. 
+ 
+ - `externaDnsHostname`: Used in conjunction with the external DNS add-on to handle automatic registration of the service with Route53.
```
Comment (kcoleman731) [inline docs/addons/nginx.md:55]: add link to addon

File: docs/addons/nginx.md [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+If [External DNS Add-on](../addons/external-dns.md) is installed, it is possible to configure NGINX ingress with an external NLB load balancer and leverage wild-card DNS domains (and public certificate) to route external traffic to individual workloads. 
+
+The following example provides support for AWS Load Balancer controller, External DNS and NGINX add-ons to enable such routing:
+
+```typescript
+const wildcardDomain = true;
+const addOns: Array<ssp.ClusterAddOn> = [
+            new ssp.AwsLoadBalancerControllerAddOn,
+            new ssp.addons.ExternalDnsAddon({
+                hostedZone: new ssp.addons.DelegatingHostedZoneProvider(
```
Comment (kcoleman731) [inline docs/addons/nginx.md:70]: consider following the `props` pattern similar to all other addon constructors?

File: docs/addons/external-dns.md [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+## Usage
+
+```typescript
+import * as ssp from '@shapirov/cdk-eks-blueprint';
+
+readonly externalDns = new ssp.addons.ExternalDnsAddon({
+    hostedZone: new ssp.addons.DelegatingHostedZoneProvider(
+        parentDomain,
+        subdomain, 
+        parentDnsAccountId,
```
Comment (kcoleman731) [inline docs/addons/external-dns.md:16]: does this addon require a parentDnsAccountId? What about hosted zones with the parentDomain in the same account?

File: examples/blueprint-construct/index.ts [hunk 13]
```diff
@@ -25,7 +25,15 @@ export default class BlueprintConstruct extends cdk.Construct {
         // AddOns for the cluster.
         const addOns: Array<ssp.ClusterAddOn> = [
             new ssp.addons.AppMeshAddOn,
-            new ssp.addons.NginxAddOn,
+            new ssp.addons.NginxAddOn({values: {
```
Comment (kcoleman731) [inline examples/blueprint-construct/index.ts:28]: is this needed?

File: lib/addons/external-dns/index.ts [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+     * @default `external-dns`
+     */
+    readonly namespace?: string;
+
+    /**
+     * @default `5.1.3`
+     */
+    readonly version?: string;
+
+    readonly hostedZone: HostedZoneProvider;
```
Comment (kcoleman731) [inline lib/addons/external-dns/index.ts:23]: doc bloc

File: lib/addons/external-dns/index.ts [hunk 15]
```diff
@@ -0,0 +1,176 @@
...
+}
+
+
+export interface HostedZoneProvider {
```
Comment (kcoleman731) [inline lib/addons/external-dns/index.ts:27]: consider moving this interface and the convenience methods to their own file.

File: lib/addons/external-dns/index.ts [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+/**
+ * Implementation of the External DNS service: https://github.com/kubernetes-sigs/external-dns/.
+ * It is required to integrate with Route53 for external DNS resolution. 
+ */
+export class ExternalDnsAddon implements ClusterAddOn {
+
+    readonly name = 'external-dns';
+
+    constructor(private readonly options: ExternalDnsProps) {
+    }
```
Comment (kcoleman731) [inline lib/addons/external-dns/index.ts:116]: nothing happening in the constructor?

**Review by shapirov103 — COMMENTED** [review]
Addressed review comments

**Review by kcoleman731 — APPROVED** [review]

**Pull request merged**